source/particles/particle_handler.cc

```diff
@@ -1118,101 +1118,84 @@ namespace Particles
+                          });
           }
+          break;
 
-        unsigned int *ndata = static_cast<unsigned int *>(data);
-        *ndata              = n_particles;
+        case parallel::distributed::Triangulation<dim, spacedim>::CELL_COARSEN:
+          // If this cell is the parent of children that will be coarsened,
+          // collect the particles of all children.
+          {
+            for (unsigned int child_index = 0;
+                 child_index < GeometryInfo<dim>::max_children_per_cell;
+                 ++child_index)
+              {
+                const typename Triangulation<dim, spacedim>::cell_iterator
+                  child = cell->child(child_index);
+                n_particles += n_particles_in_cell(child);
+              }
 
-        data = static_cast<void *>(ndata + 1);
+            stored_particles_on_cell.reserve(n_particles);
 
-        for (unsigned int child_index = 0;
-             child_index < GeometryInfo<dim>::max_children_per_cell;
-             ++child_index)
-          {
-            const typename Triangulation<dim, spacedim>::cell_iterator child =
-              cell->child(child_index);
-            const boost::iterator_range<particle_iterator> particle_range =
-              particles_in_cell(child);
-
-            for (particle_iterator particle = particle_range.begin();
-                 particle != particle_range.end();
-                 ++particle)
+            for (unsigned int child_index = 0;
+                 child_index < GeometryInfo<dim>::max_children_per_cell;
+                 ++child_index)
               {
-                particle->write_data(data);
+                const typename Triangulation<dim, spacedim>::cell_iterator
+                                         child       = cell->child(child_index);
+                const internal::LevelInd level_index = {child->level(),
+                                                        child->index()};
+                const auto               particles_in_cell =
+                  (child->is_ghost() ?
+                     ghost_particles.equal_range(level_index) :
+                     particles.equal_range(level_index));
+
+                std::for_each(particles_in_cell.first,
+                              particles_in_cell.second,
+                              [&stored_particles_on_cell](
+                                std::pair<internal::LevelInd,
+                                          Particle<dim, spacedim>> particle) {
+                                stored_particles_on_cell.push_back(
+                                  particle.second);
+                              });
               }
           }
+          break;
+
+        default:
+          Assert(false, ExcInternalError());
+          break;
       }
-    else
-      Assert(false, ExcInternalError());
 
-    return data_vector;
+    Assert(n_particles == stored_particles_on_cell.size(), ExcInternalError());
```

Comment:
move this assertion right after the two places where you call `std::foreach` to copy the particles. If you do this, then you can also make the `n_particles` variable a local variable in the two code blocks where you use it, rather than making it a function-level variable.